shotover/src/server.rs

```diff
@@ -255,6 +262,132 @@ pub struct Handler<C: CodecBuilder> {
     _permit: OwnedSemaphorePermit,
 }
 
+fn spawn_websocket_read_write_tasks<C: CodecBuilder + 'static>(
+    codec: C,
+    ws_stream: WebSocketStream<TcpStream>,
+    in_tx: UnboundedSender<Messages>,
+    mut out_rx: UnboundedReceiver<Messages>,
+    out_tx: UnboundedSender<Messages>,
+) {
+    let (mut writer, mut reader) = ws_stream.split();
+    let (mut decoder, mut encoder) = codec.build();
+
+    // read task
+    tokio::spawn(async move {
+        loop {
+            tokio::select! {
+                result = reader.next() => {
+                    if let Some(ws_message) = result {
+                        match ws_message {
+                            Ok(WsMessage::Binary(ws_message_data)) => {
+                                let message = decoder.decode(&mut BytesMut::from(ws_message_data.as_slice()));
```

Comment:
This an unfortunate clone, is there anything we can do about it?

Reply:
Hmmm... The only way this could happen is one of:
* we fork tungstenite to use a BytesMut instead of a Vec.
* BytesMut one day decides to make public its from_vec method.

For now lets just add a comment.
```
// Entire message is reallocated and copied here due to incompatibility between tokio codecs and tungstenite.
```